pkg/kubectl/cmd/convert.go

```diff
@@ -0,0 +1,159 @@
+func NewCmdConvert(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &ConvertOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "convert -f FILENAME",
+		Short:   "Convert config files between different API versions",
+		Long:    convert_long,
+		Example: convert_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := options.Complete(f, out, cmd, args)
+			cmdutil.CheckErr(err)
+			err = options.RunConvert()
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	usage := "Filename, directory, or URL to file to need to get converted."
+	kubectl.AddJsonFilenameFlag(cmd, &options.filenames, usage)
+	cmd.MarkFlagRequired("filename")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().BoolVar(&options.local, "local", false, "If true, convert will NOT try to contact api-server but run locally.")
+
+	return cmd
+}
+
+// ConvertOptions have the data required to perform the convert operation
+type ConvertOptions struct {
+	builder   *resource.Builder
+	filenames []string
+	local     bool
+
+	out     io.Writer
+	printer kubectl.ResourcePrinter
+
+	outputVersion string
+}
+
+// Complete collects information required to run Convert command from command line.
+func (o *ConvertOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) (err error) {
+	o.outputVersion = cmdutil.OutputVersion(cmd, latest.GroupOrDie("").Version)
+	if !registered.IsRegisteredAPIVersion(o.outputVersion) {
+		cmdutil.UsageError(cmd, "'%s' is not a registered version.", o.outputVersion)
+	}
+
+	// build the builder
+	mapper, typer := f.Object()
+	o.builder = resource.NewBuilder(mapper, typer, f.ClientMapperForCommand())
+	enforceNamespace := false
+	if o.local {
+		fmt.Fprintln(out, "running in local mode...")
+		o.builder = o.builder.Local()
+	} else {
+		schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
+		if err != nil {
+			return err
+		}
+		cmdNamespace, enforce, err := f.DefaultNamespace()
+		enforceNamespace = enforce
+		if err != nil {
+			return err
+		}
+		o.builder = o.builder.Schema(schema).
+			NamespaceParam(cmdNamespace).DefaultNamespace()
+	}
+	o.builder = o.builder.ContinueOnError().
+		FilenameParam(enforceNamespace, o.filenames...).
+		Flatten()
+
+	// build the printer
+	outputFormat := cmdutil.GetFlagString(cmd, "output")
+	templateFile := cmdutil.GetFlagString(cmd, "template")
+	if len(outputFormat) == 0 {
+		if len(templateFile) == 0 {
+			outputFormat = "yaml"
+		} else {
+			outputFormat = "template"
+		}
+	}
+
+	o.printer, _, err = kubectl.GetPrinter(outputFormat, templateFile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RunConvert implements the generic Convert command
+func (o *ConvertOptions) RunConvert() error {
+
+	infos, err := o.builder.Do().Infos()
+	if err != nil {
+		return err
+	}
+
+	objects, err := resource.AsVersionedObject(infos, false, o.outputVersion)
```

Comment:
> I think with AsVersionedObject only one list will be returned. And kind info is also kept for every item. So both are fine. Only because AsVersionedObject doesn't need extra loop to display it, so I choose to use this one.

Do `Lists` contain `Lists` properly?  I haven't checked, but if I tried to convert an entire directory, I'd be a little tweaked if my lists of elements weren't preserved as a distinct list inside of the list.

To clarify what I mean.  `dir/one-resource.yaml` and `dir/list-of-resources.yaml`.  When I call `kubectl convert dir`, I want 
```
List
    one-resource
    List
        list-item1
        list-item2
```

**not**
```
List
    one-resource
    list-item1
    list-item2
```